devicetypes/smartthings/samsung-smart-doorlock.src/samsung-smart-doorlock.groovy

```diff
@@ -0,0 +1,402 @@
+	log.trace "ZigBee DTH - Executing refresh() for device ${device.displayName}"
+	def cmds = zigbee.readAttribute(CLUSTER_DOORLOCK, DOORLOCK_ATTR_DOORSTATE)
+	log.info "ZigBee DTH - refresh() returning with cmds:- $cmds"
+	return cmds
+}
+
+
+/**
+ * Configures the device to settings needed by SmarthThings at device discovery time
+ *
+ */
+def configure() {
+	log.trace "ZigBee DTH - Executing configure() for device ${device.displayName}"
+	def cmds = zigbee.command(0x0000, 0x1E, "",[mfgCode: 0003]) // read modelName of the device as it is not being sent with zbjoin.
+    state.configured = true
+    sendEvent(name: "lock", value: "unlocked", isStateChange: true, displayed: false)
+    sendEvent(name: "battery", value: "100", isStateChange: true, displayed: false)
+  	// Device-Watch allows 2 check-in misses from device + ping (plus 2 min lag time)
+	sendEvent(name: "checkInterval", value: 2 * 60 * 60 + 2 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID, offlinePingable: "1"])
+
+	log.info "ZigBee DTH - configure() returning with cmds:- $cmds"
+	cmds
+}
+
+
+/**
+ * Executes unlock command on a Zigbee lock
+ */
+def unlock() {
+	log.trace "ZigBee DTH - Executing unlock() for device ${device.displayName}"
+	def cmds = zigbee.command(CLUSTER_DOORLOCK, DOORLOCK_CMD_UNLOCK_DOOR, "100431323335", [mfgCode: 0003])
+    log.info "ZigBee DTH - unlock() returning with cmds:- $cmds"
+	return cmds
+}
+
+
+/**
+ * Responsible for parsing incoming device messages to generate events
+ *
+ * @param description The incoming description from the device
+ *
+ * @return result: The list of events to be sent out
+ *
+ */
+def parse(String description) {
+	log.trace "ZigBee DTH - Executing parse() for device ${device.displayName}"
+	def result = null
+	if (description) {
+		if (description.startsWith('read attr -')) {
+			result = parseAttributeResponse(description)
+		} else {
+			result = parseCommandResponse(description)
+		}
+	}
+	return result
+}
+
+/**
+ * Responsible for handling attribute responses
+ *
+ * @param description The description to be parsed
+ *
+ * @return result: The list of events to be sent out
+ */
+private def parseAttributeResponse(String description) {
+	Map descMap = zigbee.parseDescriptionAsMap(description)
+	log.trace "ZigBee DTH - Executing parseAttributeResponse() for device ${device.displayName} with description map:- $descMap"
+	def result = []
+	Map responseMap = [:]
+	def clusterInt = descMap.clusterInt
+	def attrInt = descMap.attrInt
+	def deviceName = device.displayName
+	if (clusterInt == CLUSTER_POWER && attrInt == POWER_ATTR_BATTERY_VOLTAGE) {
+		responseMap.name = "battery"
+		responseMap.value = getBatteryResult(Integer.parseInt(descMap.value, 10))
+		responseMap.descriptionText = "Battery is at ${responseMap.value}%"
+	} else if (clusterInt == CLUSTER_DOORLOCK && attrInt == DOORLOCK_ATTR_LOCKSTATE) {
+		def value = Integer.parseInt(descMap.value, 16)
+		responseMap.name = "lock"
```

Comment:
Is this code needed since operational events are also being handled?

Reply:
Can you please provide little more explanation on which part of code is unnecessary? Thank you.

Reply:
It doesn't appear necessary to generate lock events on DoorLock Attribute Reports since Operational Events are also being sent and the resulting event contains more information like the source of the event. As it stands it appears two lock or unlock events would be created each time the state changes - one because of the attribute report and one because of the operational event.

Reply:
Line 279 where the cmd is `DOORLOCK_RESPONSE_OPERATION_EVENT`

Reply:
When parsing the event, wouldn't this be considered two different events? One is to parse the result of command from ST app, and the other one is to parse the attribute report or any activity of the doorlock that occurs outside the ST app.

Reply:
I guess the question is if locking and unlocking the door is generating two lock events or just one? The code looks like it will generate two, one with the `method` and one without.